### ISSUE TITLE
guarantees for the logs argument in keras Callbacks

### ISSUE THREAD

**Issue by ngc92** [tensorflow/tensorflow#37187]
## URL(s) with the issue:
https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/Callback
https://www.tensorflow.org/guide/keras/custom_callback

## Description of issue (what needs changing):
The documentation for the keras `Callback` base class contains the following generic statement about the `logs` parameter passed to its methods:
```
The logs dictionary that callback methods take as argument will contain keys for quantities relevant to the current batch or epoch.
```
and
```
The logs dict contains the loss value, and all the metrics at the end of a batch or epoch. Example includes the loss and mean absolute error.
```
on the Keras custom callbacks page.

Since python passes objects by reference, the question becomes whether write-access to this logs parameter is allowed and supported. An example use case would be to provide a custom callback that populates the `logs` dictionary with some additional information that than would automatically be displayed in the progress bar and tensorboard, and recorded in history and CSV callbacks. 

Therefore, I think the documentation should clearly state whether 
1) Write-access to the `logs` dict is forbidden (in which case it might be worthwhile to pass a non-writeable dict-like type)
2) Write-access to `logs` is allowed, and will not have any side-effects on any other Callback (i.e. each Callback gets an independent copy)
3) The `logs` dict is writable, and changes to it are visible to any further Callback. This would also require to specify in which order the callbacks are processed.

**Comment by CMCDragonkai** [comment]
I'm interested in clarifying this as well. Since I currently assign a new metric into the logs dictionary, and the metric ends up being reported in tensorboard, but not on the terminal even with `verbose=1`.

It would nice to be able to add whatever metric we want into the tensorboard.

**Comment by adriang133** [comment]
+1

Very confusing that updates to the logs dict inside a `keras.Callback` do not make it to the terminal.

**Comment by CubasMike** [comment]
Seems this issue is still not solved. Here we are redirected to #44940, and there we are redirected back here. I have the same issue. Can someone please clarify how the callbacks "logs" parameters behave?

I cannot migrate from TF 2.0 to TF 2.2 and this is one of the issues.

**Comment by NijatR** [comment]
I came across this issue while trying to pass a custom metric to `ModelCheckpoint` callback from a custom callback. `logs` dictionary can be used for this purpose, however `_supports_tf_logs` flag of the `ModelCheckpoint` instance needed to be set to `False`, otherwise the `logs` dictionary was not updated and passed to it. Additionally, what I observed is that the callbacks are executed in the order they are defined and `logs` dictionary is updated in that order as well.

Sample code:
```
# Define the custom callback for the metric
class CustomMetricCallback(tf.keras.callbacks.Callback):
    def __init__(self, logs={}):
        super().__init__()

    def on_epoch_end(self, epoch, logs={}):
        logs['customer_metric_name'] = ... # compute the metric value
```
```
# Initialize the checkpoint callback with the custom metric
weights_path = 'checkpoint-{epoch:02d}-{customer_metric_name:.4f}.h5'
checkpoint = tf.keras.callbacks.ModelCheckpoint(weights_path, monitor='customer_metric_name')
checkpoint._supports_tf_logs = False
```
```
# Define the model callbacks
callbacks = [
    CustomMetricCallback(), # should be listed before the other callbacks that will use its logs dictionary
    checkpoint
]
model.fit(train_gen, validation_data=val_gen, callbacks=callbacks,)
```

Tested with tensorflow version 2.4.1.

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
